ports/rp2/mphalport.c

```diff
@@ -273,21 +286,39 @@ uint32_t storage_read_blocks(uint8_t *dest, uint32_t block_num, uint32_t num_blo
     panic_unsupported();
 }
 
-static int64_t soft_timer_callback(alarm_id_t id, void *user_data) {
-    soft_timer_alarm_id = 0;
-    pendsv_schedule_dispatch(PENDSV_DISPATCH_SOFT_TIMER, soft_timer_handler);
-    return 0; // don't reschedule this alarm
-}
-
 uint32_t soft_timer_get_ms(void) {
     return mp_hal_ticks_ms();
 }
 
 void soft_timer_schedule_at_ms(uint32_t ticks_ms) {
-    if (soft_timer_alarm_id != 0) {
-        cancel_alarm(soft_timer_alarm_id);
-    }
     int32_t ms = soft_timer_ticks_diff(ticks_ms, mp_hal_ticks_ms());
     ms = MAX(0, ms);
-    soft_timer_alarm_id = add_alarm_in_ms(ms, soft_timer_callback, NULL, true);
+    if (hardware_alarm_set_target(MICROPY_HW_SOFT_TIMER_ALARM_NUM, delayed_by_ms(get_absolute_time(), ms))) {
+        // "missed" hardware alarm target
+        hardware_alarm_force_irq(MICROPY_HW_SOFT_TIMER_ALARM_NUM);
+    }
+}
+
+void soft_timer_init(void) {
+    hardware_alarm_claim(MICROPY_HW_SOFT_TIMER_ALARM_NUM);
+    hardware_alarm_set_callback(MICROPY_HW_SOFT_TIMER_ALARM_NUM, (void *)soft_timer_handler);
+    // soft_timer_handler must be executed at IRQ_PRI_PENDSV priority. Rather
+    // than having the timer IRQ trigger the handler via a PendSV interrupt,
+    // have the timer IRQ execute it directly at the correct priority.
+    NVIC_SetPriority(TIMER_IRQ_0_IRQn + MICROPY_HW_SOFT_TIMER_ALARM_NUM, IRQ_PRI_PENDSV);
+}
+
+void mp_wfe_or_timeout(uint32_t timeout_ms) {
+    soft_timer_entry_t timer;
+
+    soft_timer_static_init(&timer, SOFT_TIMER_MODE_ONE_SHOT, 0, NULL);
+
+    // Note the timer doesn't have an associated callback, it just exists to create a
+    // hardware interrupt to wake the CPU
+    soft_timer_insert(&timer, timeout_ms);
```

Comment:
Did you intend to reorganise these lines as I suggested before?

Reply:
Um, sorry no - did I miss a comment somewhere?

Reply:
Oh, I definitely wrote the comment and was sure I saw it after posting... anyway...

I suggest:
```c
    // Note the timer doesn't have an associated callback, it just exists to create a
    // hardware interrupt to wake the CPU
    soft_timer_static_init(...);
    soft_timer_insert(&timer, timeout_ms);
```

Reply:
Done!